_posts/2016-05-13-rustup.md

```diff
@@ -0,0 +1,553 @@
+```
+rust:~$ cargo new --bin hello && cd hello
+rust:~/hello$ rustup target add arm-linux-androideabi
+info: downloading component 'rust-std' for 'arm-linux-androideabi'
+info: installing component 'rust-std' for 'arm-linux-androideabi'
+rust:~/hello$ rustup show
+installed targets for active toolchain
+--------------------------------------
+
+arm-linux-androideabi
+x86_64-unknown-linux-gnu
+
+active toolchain
+----------------
+
+stable-x86_64-unknown-linux-gnu (default)
+rustc 1.8.0 (db2939409 2016-04-11)
+```
+
+So let's see what happens if we try to just build our 'hello'
+project without installing anything further:
+
+```
+rust:~/hello$ cargo build --target=arm-linux-androideabi
+   Compiling hello v0.1.0 (file:///home/rust/hello)
+error: linking with `cc` failed: exit code: 1
+... (lots of noise elided)
+error: aborting due to previous error
+Could not compile `hello`.
+```
+
+The problem is that we don't have a linker that supports Android yet,
+so let's take a moment's digression to talk about building for
+Android. To develop for Android we need the [Andoid NDK]. It contains
+the linker `rustc` needs to create Android binaries. To just *build*
+Rust code that targets Android the only thing we need is the NDK, but
+for practical development we'll want the [Android SDK] too.
+
+[Android NDK]: https://developer.android.com/ndk/guides/setup.html#install
+[Android SDK]: https://developer.android.com/sdk/index.html
+
+On Linux, download and unpack them with the following commands (the
+output of which is not included here):
+
+```
+rust:~/home$ cd
+rust:~$ curl -O https://dl.google.com/android/android-sdk_r24.4.1-linux.tgz
+rust:~$ tar xzf android-sdk_r24.4.1-linux.tgz
+rust:~$ curl -O http://dl.google.com/android/repository/android-ndk-r10e-linux-x86_64.zip
+rust:~$ unzip android-ndk-r10e-linux-x86_64.zip
+```
+
+We further need to create what the NDK calls a ["standalone
+toolchain"].  We're going to put ours in a directory called
+`android-ndk-r10e`:
+
+["standalone toolchain"]: https://developer.android.com/ndk/guides/standalone_toolchain.html
+
+```
+rust:~$ android-ndk-r10e/build/tools/make-standalone-toolchain.sh \
+      --platform=android-18 --toolchain=arm-linux-androideabi-clang3.6 \
+      --install-dir=android-18-toolchain --ndk-dir=android-ndk-r10e/ --arch=arm
+Auto-config: --toolchain=arm-linux-androideabi-4.8, --llvm-version=3.6
+Copying prebuilt binaries...
+Copying sysroot headers and libraries...
+Copying c++ runtime headers and libraries...
+Copying files to: android-18-toolchain
+Cleaning up...
+Done.
+```
+
+Let's notice a few things about these commands. First, the NDK we
+downloaded, `android-ndk-r10e-linux-x86_64.zip` is not the most recent
+release (which at the time of this writing is 'r11c'). Rust's `std` is
+built against `r10e` and links to symbols that are no longer included
+in the NDK. So for now we have to use the older NDK.  Second, in
+building the standalone toolchain we passed "--platform=android-18" to
```

Comment:
might want to make this backticks instead of double quotes so it renders like code